src/Microsoft.AspNet.OData.Shared/ODataAPIHandler.cs

```diff
@@ -0,0 +1,186 @@
+            PropertyInfo property = type.GetProperties().FirstOrDefault(s => s.PropertyType == typeof(ODataIdContainer));
+            if (property != null && property.GetValue(obj) is ODataIdContainer container && container != null)
+            {
+                ODataPath odataPath = GetODataPath(container.ODataId, model);
+                object res = ApplyODataIdOnContainer(obj, odataPath);
+                foreach (PropertyInfo prop in type.GetProperties())
+                {
+                    object resVal = prop.GetValue(res);
+
+                    if (resVal != null)
+                    {
+                        prop.SetValue(obj, resVal);
+                    }
+                }
+            }
+            else
+            {
+                foreach (PropertyInfo prop in type.GetProperties().Where(p => !p.PropertyType.IsPrimitive))
+                {
+                    object propVal = prop.GetValue(obj);
+                    if (propVal == null)
+                    {
+                        continue;
+                    }
+
+                    if (propVal is IEnumerable lst)
+                    {
+                        foreach (object item in lst)
+                        {
+                            if (item.GetType().IsPrimitive)
+                            {
+                                break;
+                            }
+
+                            CheckAndApplyODataId(item, model);
+                        }
+                    }
+                    else
+                    {
+                        CheckAndApplyODataId(propVal, model);
+                    }
+                }
+            }
+        }
+
+        private object ApplyODataIdOnContainer(object obj, ODataPath odataPath)
+        {
+            KeySegment keySegment = odataPath.LastOrDefault() as KeySegment;
+            Dictionary<string, object> keys = keySegment?.Keys.ToDictionary(x => x.Key, x => x.Value);
+            if (keys != null)
+            {
+                TStructuralType returnedObject;
+                string error;
+                if (this.ParentObject.Equals(obj))
+                {
+                    if (this.TryGet(keys, out returnedObject, out error) == ODataAPIResponseStatus.Success)
+                    {
+                        return returnedObject;
+                    }
+                    else
+                    {
+                        if (this.TryCreate(keys, out returnedObject, out error) == ODataAPIResponseStatus.Success)
+                        {
+                            return returnedObject;
+                        }
+                        else
+                        {
+                            return null;
+                        }
+                    }
+                }
+                else
+                {
+                    IODataAPIHandler apiHandlerNested = this.GetNestedHandler(this.ParentObject, odataPath.LastSegment.Identifier);
```

Comment:
Why didn't you pass `ODataPath` to this method? That means the whole logic will go..Also if we do not to provide this deep insert logic we can get rid of it entirely from this class.

Reply:
I updated the `UpdateLinkedObjects` method to `internal`. The other related methods are `private`. We can leave them as they are and clean up once we redesign deep insert.